refactor(x/mint, x/gamm errors): move errors in mint and gamm keeper to errors.go

stackman27 opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
v    If your PR doesn't close an issue, that's OK!  Just remove the Closes: #XXX line!
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #2034 and #2035

## What is the purpose of the change
Move tests to errors.go for consistency 

## Brief Changelog
n/a

## Testing and Verifying
n/a

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

stackman27 commented on x/mint/keeper/keeper.go:
```diff
@@ -71,7 +62,7 @@ func NewKeeper(
 func (k Keeper) SetInitialSupplyOffsetDuringMigration(ctx sdk.Context) error {
 	if !k.accountKeeper.HasAccount(ctx, k.accountKeeper.GetModuleAddress(types.DeveloperVestingModuleAcctName)) {
-		return errDevVestingModuleAccountNotCreated
+		return types.ErrDevVestingModuleAccountNotCreated
```
I thought about wrapping this with sdkerrors like; 
```
sdkerrors.Wrapf(types.ErrDevVestingModuleAccountNotCreated, "%s module account already exists", types.DeveloperVestingModuleAcctName")
```

But then i'd have to change the way we're catching errors in keeper_test? Any suggestions 

cc @p0mvn

p0mvn replied:
Is refactoring to something like the following possible?
https://github.com/osmosis-labs/osmosis/blob/04424733e90c7b6f66264a9db6d71b48cb62ddba/x/mint/keeper/keeper_test.go#L636

Where the expected error is an sdk wrapper?

stackman27 replied:
1. I haven't seen sdkerrors used in any test at all, so wouldn't it be better to maintain consistency? 
2. Also, comparing sdkerrors in test seems very confusing, for instance this output is shown unequal 
```
 Not equal: 
 expected: vesting module account already exist: module account already exists [/Users/sishirg27/Desktop/osmosis/x/mint/keeper/keeper.go:86]
 actual  : vesting module account already exist: module account already exists [/Users/sishirg27/Desktop/osmosis/x/mint/keeper/keeper_test.go:238]
```

stackman27 replied:
1. I haven't really seen sdkerrors being used in test elsewhere, would i be better to maintain consistency? 
2. Comparing sdkerrors is super weird, for example this is showing unequal 
```
 Not equal: 
expected: vesting module account doesnot exist: module account does not exist [/Users/sishirg27/Desktop/osmosis/x/mint/keeper/keeper.go:66]
actual  : vesting module account doesnot exist: module account does not exist [/Users/sishirg27/Desktop/osmosis/x/mint/keeper/keeper_test.go:273]
                                    
    Diff:
    --- Expected
    +++ Actual
    @@ -8,7 +8,8 @@
       }),
    -  stack: (*errors.stack)((len=6) {
    +  stack: (*errors.stack)((len=7) {
        (uintptr) 0x4625d7b,
    -   (uintptr) 0x5061f88,
    -   (uintptr) 0x552c858,
    -   (uintptr) 0x5068896,
    +   (uintptr) 0x552c436,
    +   (uintptr) 0x40d1865,
    +   (uintptr) 0x40d0ddf,
    +   (uintptr) 0x5069656,
        (uintptr) 0x411a322,
```

p0mvn replied:
1. I think that's because we haven't been asserting the kinds of errors in tests in general. It is preferable to expect a specific error in tests to ensure we cover the desired branches in the tests.
2. This might work: `suite.ErrorIs(actualError, tc.expectedError)`

Let me know what you think

stackman27 replied:
THAT WORKED!!

AlpinYukseloglu commented on x/mint/keeper/keeper_test.go:
```diff
@@ -361,7 +361,7 @@ func (suite *KeeperTestSuite) TestSetInitialSupplyOffsetDuringMigration() {
 		"dev vesting module account does not exist": {
 			blockHeight:   1,
-			expectedError: keeper.ErrDevVestingModuleAccountNotCreated,
+			expectedError: sdkerrors.Wrapf(types.ErrModuleDoesnotExist, "%s vesting module account doesnot exist", types.DeveloperVestingModuleAcctName),
```
```suggestion
			expectedError: sdkerrors.Wrapf(types.ErrModuleDoesnotExist, "%s vesting module account does not exist", types.DeveloperVestingModuleAcctName),
```

AlpinYukseloglu commented on x/mint/keeper/keeper_test.go:
```diff
@@ -340,7 +340,7 @@ func (suite *KeeperTestSuite) TestCreateDeveloperVestingModuleAccount() {
 			if tc.expectedError != nil {
 				suite.Error(actualError)
-				suite.Equal(actualError, tc.expectedError)
+				suite.ErrorIs(actualError, tc.expectedError)
```
Is this used anywhere else in our codebase?

Also, I believe `suite.ErrorIs` just makes sure that the expected error is a _subset_ of the actual error (see testify implementation [here](https://github.com/stretchr/testify/blob/181cea6eab8b2de7071383eca4be32a424db38dd/assert/assertion_forward.go#L249)), whereas `suite.Equal` makes sure both are exactly equal, so I have a light preference for keeping this as `suite.Equal`.

alexanderbez replied:
You want to use `errors.Is` (https://pkg.go.dev/errors#Is), so you can safely catch and compare the error, even if it's wrapped.

stackman27 replied:
Had a chat with @p0mvn about this and decided to go with `suite.ErrorIs` because it was very difficult to compare sdkerrors using `suite.Equal`. For instance this is showing unequal;
```
Not equal: 
expected: vesting module account doesnot exist: module account does not exist 
actual  : vesting module account doesnot exist: module account does not exist
                                    
    Diff:
    --- Expected
    +++ Actual
    @@ -8,7 +8,8 @@
       }),
    -  stack: (*errors.stack)((len=6) {
    +  stack: (*errors.stack)((len=7) {
        (uintptr) 0x4625d7b,
    -   (uintptr) 0x5061f88,
    -   (uintptr) 0x552c858,
    -   (uintptr) 0x5068896,
    +   (uintptr) 0x552c436,
    +   (uintptr) 0x40d1865,
    +   (uintptr) 0x40d0ddf,
    +   (uintptr) 0x5069656,
        (uintptr) 0x411a322,
```

p0mvn replied:
`suite.Equal()` would compare error addresses which are not equal, `suite.ErrorIs()` is desirable here

AlpinYukseloglu reviewed (approved):
Left a couple comments for small preferences/changes but overall LGTM!

p0mvn commented on x/gamm/pool-models/balancer/pool.go:
```diff
@@ -17,14 +17,15 @@ import (
+	invalidInputDenomsErrFormat     = "input denoms must already exist in the pool (%s)"
+
+	// the errors below are being used in export_test
 	errMsgFormatFailedInterimLiquidityUpdate  = "failed to update interim liquidity - pool asset %s does not exist"
 	errMsgFormatRepeatingPoolAssetsNotAllowed = "repeating pool assets not allowed, found %s"
 	errMsgFormatNoPoolAssetFound              = "can't find the PoolAsset (%s)"
```
The top messages were changed to be suffixed with `ErrFormat`. Can we change these as well please for consistency?

alexanderbez replied:
good idea!

p0mvn reviewed (approved):
LGTM

mattverse commented on x/gamm/pool-models/balancer/pool.go:
```diff
@@ -17,14 +17,15 @@ import (
+	sharesLargerThanMaxErrFormat    = "%d resulted shares is larger than the max amount of %d"
+	invalidInputDenomsErrFormat     = "input denoms must already exist in the pool (%s)"
+
+	// the errors below are being used in export_test
```
I think this comment is misleading?(called in `amm_test.go`)

stackman27 replied:
removed

mattverse commented on x/mint/types/errors.go:
```diff
@@ -0,0 +1,11 @@
+
+var (
+	ErrAmountNilOrZero    = sdkerrors.Register(ModuleName, 2, "amount cannot be nil or zero")
+	ErrModuleAlreadyExist = sdkerrors.Register(ModuleName, 3, "module account already exists")
```
```suggestion
	ErrModuleAccountAlreadyExist = sdkerrors.Register(ModuleName, 3, "module account already exists")
```

mattverse commented on x/mint/keeper/keeper.go:
```diff
@@ -96,10 +90,10 @@ func (k Keeper) SetInitialSupplyOffsetDuringMigration(ctx sdk.Context) error {
 	}
 	if k.accountKeeper.HasAccount(ctx, k.accountKeeper.GetModuleAddress(types.DeveloperVestingModuleAcctName)) {
-		return errDevVestingModuleAccountAlreadyCreated
+		return sdkerrors.Wrapf(types.ErrModuleAlreadyExist, "%s vesting module account already exist", types.DeveloperVestingModuleAcctName)
```
```suggestion
		return sdkerrors.Wrapf(types.ErrModuleAccountAlreadyExist, "%s vesting module account already exist", types.DeveloperVestingModuleAcctName)
```

mattverse reviewed (approved):
LGTM!

stackman27 commented:
Thank you everyone for the feedback! Addressed all the comments here [`e62774b`](https://github.com/osmosis-labs/osmosis/pull/2067/commits/e62774bc3ffc1040ff59043194f95d3c1f7cef90)

p0mvn reviewed (approved):
Great work :rocket:
